File: Java/Dynamic-Programming/Fractional_knapsack/Fractional_knapsack.java
import java.util.*;

class Item {
    int weight;
    int value;

    public Item(int weight, int value) {
        this.weight = weight;
        this.value = value;
    }
}

class KnapsackComparator implements Comparator<Item> {
    public int compare(Item item1, Item item2) {
        double ratio1 = (double) item1.value / item1.weight;
        double ratio2 = (double) item2.value / item2.weight;
        if (ratio1 < ratio2)
            return 1;
        else if (ratio1 > ratio2)
            return -1;
        else
            return 0;
    }
}

public class FractionalKnapsack {
    public static double getMaxValue(Item[] items, int capacity) {
        Arrays.sort(items, new KnapsackComparator());

        double totalValue = 0;
        int currentWeight = 0;

        for (Item item : items) {
            if (currentWeight + item.weight <= capacity) {
                currentWeight += item.weight;
                totalValue += item.value;
            } else {
                int remainingCapacity = capacity - currentWeight;
                totalValue += (double) item.value / item.weight * remainingCapacity;
                break;
            }
        }

        return totalValue;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter the number of items: ");
        int n = scanner.nextInt();

        Item[] items = new Item[n];
        for (int i = 0; i < n; i++) {
            System.out.print("Enter weight and value for item " + (i + 1) + ": ");
            int weight = scanner.nextInt();
            int value = scanner.nextInt();
            items[i] = new Item(weight, value);
        }

        System.out.print("Enter the knapsack capacity: ");
        int capacity = scanner.nextInt();

        scanner.close();

        double maxValue = getMaxValue(items, capacity);
        System.out.println("The maximum value that can be obtained is: " + maxValue);
    }
}

File: C++/Dynamic-Programming/Fractional_knapsack/Fractional_knapsack.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// Structure representing an item with weight and value
struct Item {
    int weight;
    int value;

    Item(int weight, int value) : weight(weight), value(value) {}
};

// Functor to compare items based on their value-to-weight ratio
struct KnapsackComparator {
    bool operator()(const Item& item1, const Item& item2) const {
        // Calculate the value-to-weight ratio for each item
        double ratio1 = static_cast<double>(item1.value) / item1.weight;
        double ratio2 = static_cast<double>(item2.value) / item2.weight;

        // Compare the ratios to determine the order
        if (ratio1 < ratio2) {
            return true;
        }
        return false;
    }
};

// Function to solve the Fractional Knapsack problem
double fractionalKnapsack(int capacity, vector<Item>& items) {
    // Sort the items based on their value-to-weight ratio using the KnapsackComparator
    sort(items.begin(), items.end(), KnapsackComparator());

    double totalValue = 0;
    int remainingCapacity = capacity;

    // Iterate through the sorted items
    for (const Item& item : items) {
        // If the item can be fully included
        if (item.weight <= remainingCapacity) {
            totalValue += item.value;
            remainingCapacity -= item.weight;
        } else { // If the item can only be partially included
            // Calculate the fraction of the item that can be included
            double fraction = static_cast<double>(remainingCapacity) / item.weight;
            totalValue += item.value * fraction;
            break;
        }
    }

    return totalValue;
}

int main() {
    int numItems;
    cout << "Enter the number of items: ";
    cin >> numItems;

    vector<Item> items(numItems);

    for (int i = 0; i < numItems; ++i) {
        cout << "Enter weight and value for item " << i + 1 << ": ";
        cin >> items[i].weight >> items[i].value;
    }

    int capacity;
    cout << "Enter the knapsack capacity: ";
    cin >> capacity;

    double maxValue = fractionalKnapsack(capacity, items);

    cout << "The maximum value that can be obtained is: " << maxValue << endl;

    
}


/*
example output: 
Enter the number of items: 4
Enter weight and value for item 1: 2 10
Enter weight and value for item 2: 3 15
Enter weight and value for item 3: 5 20
Enter weight and value for item 4: 1 5
Enter the knapsack capacity: 7
The maximum value that can be obtained is: 30.7143

*/

